src/main/java/org/wso2/carbon/hash/provider/pbkdf2/PBKDF2HashProvider.java

```diff
@@ -0,0 +1,213 @@
+
+import org.wso2.carbon.user.core.exceptions.HashProviderClientException;
+import org.wso2.carbon.user.core.exceptions.HashProviderException;
+import org.wso2.carbon.user.core.exceptions.HashProviderServerException;
+import org.wso2.carbon.user.core.hash.HashProvider;
+
+import java.nio.charset.StandardCharsets;
+import java.security.NoSuchAlgorithmException;
+import java.security.spec.InvalidKeySpecException;
+import java.util.Base64;
+import java.util.Map;
+import javax.crypto.SecretKeyFactory;
+import javax.crypto.spec.PBEKeySpec;
+
+/**
+ * This class contains the implementation of PBKDF2 hashing algorithm.
+ */
+public class PBKDF2HashProvider implements HashProvider {
+
+    private static final Log log = LogFactory.getLog(PBKDF2HashProvider.class);
+
+    private String pseudoRandomFunction;
+    private int dkLength;
+    private int iterationCount;
+
+    @Override
+    public void init() {
+
+        pseudoRandomFunction = Constants.DEFAULT_PBKDF2_PRF;
+        dkLength = Constants.DEFAULT_DERIVED_KEY_LENGTH;
+        iterationCount = Constants.DEFAULT_ITERATION_COUNT;
+    }
+
+    @Override
+    public void init(Map<String, Object> initProperties) {
+
+        iterationCount = resolveIterationCount(initProperties);
+        dkLength = resolveDerivedKeyLength(initProperties);
+        pseudoRandomFunction = resolvePseudoRandomFunction(initProperties);
+    }
+
+    @Override
+    public byte[] getHash(String value, String salt) throws HashProviderException {
+
+        validateEmptyValue(value);
+        validateSalt(salt);
+        return calculateHash(value, salt, iterationCount, dkLength, pseudoRandomFunction);
+    }
+
+    @Override
+    public byte[] getHash(String value, String salt, Map<String, Object> metaProperties)
+            throws HashProviderException {
+
+        validateEmptyValue(value);
+        validateSalt(salt);
+        int iterations = resolveIterationCount(metaProperties);
+        int dkLen = resolveDerivedKeyLength(metaProperties);
+        String prf = resolvePseudoRandomFunction(metaProperties);
+        return calculateHash(value, salt, iterations, dkLen, prf);
+    }
+
+    @Override
+    public String getAlgorithm() {
+
+        return Constants.PBKDF2_HASHING_ALGORITHM;
+    }
+
+    /**
+     * Calculate hash value according to the given parameters.
+     *
+     * @param value                Value to be hashed.
+     * @param salt                 The salt.
+     * @param iterationCount       Number of iterations to be used by the PRF.
+     * @param dkLength             The output length of the hash function.
+     * @param pseudoRandomFunction PRF function which needs to be used for PBKDF2 hashing.
+     * @return The resulting hash value of the value.
+     * @throws HashProviderException If an error occurred while calculating the hash.
+     */
+    private byte[] calculateHash(String value, String salt, int iterationCount, int dkLength,
```

Comment:
Lets add `int iterationCount, int dkLength` at the end. Make sure to update the method comment as well.